src/library/scala/collection/mutable/OrderAwareHashMap.scala

```diff
@@ -0,0 +1,800 @@
+    if (nd eq null) default else nd match {
+      case nd: LLNode => nd.value
+      case n => n.asInstanceOf[RBNode].value
+    }
+  }
+
+  @`inline` private[this] def findNode(elem: K): Node = {
+    val hash = computeHash(elem)
+    table(index(hash)) match {
+      case null => null
+      case n: LLNode => n.getNode(elem, hash)
+      case n => n.asInstanceOf[RBNode].getNode(elem, hash)
+    }
+  }
+
+  override def sizeHint(size: Int): Unit = {
+    val target = tableSizeFor(((size + 1).toDouble / loadFactor).toInt)
+    if(target > table.length) {
+      if(size == 0) reallocTable(target)
+      else growTable(target)
+    }
+  }
+
+  override def update(key: K, value: V): Unit = put0(key, value, false)
+
+  override def put(key: K, value: V): Option[V] = put0(key, value, true) match {
+    case null => None
+    case sm => sm
+  }
+
+  def addOne(elem: (K, V)): this.type = { put0(elem._1, elem._2, false); this }
+
+  @`inline` private[this] def put0(key: K, value: V, getOld: Boolean): Some[V] = {
+    if(contentSize + 1 >= threshold) growTable(table.length * 2)
+    val hash = computeHash(key)
+    val idx = index(hash)
+    put0(key, value, getOld, hash, idx)
+  }
+
+  private[this] def put0(key: K, value: V, getOld: Boolean, hash: Int, idx: Int): Some[V] = {
+    val res = table(idx) match {
+      case n: RBNode =>
+        insert(n, idx, key, hash, value)
+      case _old =>
+        val old: LLNode = _old.asInstanceOf[LLNode]
+        if(old eq null) {
+          table(idx) = new LLNode(key, hash, value, null)
+        } else {
+          var remaining = OrderAwareHashMap.treeifyThreshold
+          var prev: LLNode = null
+          var n = old
+          while((n ne null) && n.hash <= hash && remaining > 0) {
+            if(n.hash == hash && key == n.key) {
+              val old = n.value
+              n.value = value
+              return (if(getOld) Some(old) else null)
+            }
+            prev = n
+            n = n.next
+            remaining -= 1
+          }
+          if(remaining == 0) {
+            treeify(old, idx)
+            return put0(key, value, getOld, hash, idx)
+          }
+          if(prev eq null) table(idx) = new LLNode(key, hash, value, old)
+          else prev.next = new LLNode(key, hash, value, prev.next)
+        }
+        true
+    }
+    if(res) contentSize += 1
+    if(res) Some(null.asInstanceOf[V]) else null //TODO
+  }
+
+  private[this] def treeify(old: LLNode, idx: Int): Unit = {
+    table(idx) = OrderAwareHashMap.leaf(old.key, old.hash, old.value, red = false, null)
+    var n: LLNode = old.next
+    while(n ne null) {
+      val root = table(idx).asInstanceOf[RBNode]
+      insertIntoExisting(root, idx, n.key, n.hash, n.value, root)
+      n = n.next
+    }
+  }
+
+  override def addAll(xs: IterableOnce[(K, V)]): this.type = {
+    val k = xs.knownSize
+    if(k > 0) sizeHint(contentSize + k)
+    super.addAll(xs)
+  }
+
+  private[this] def remove0(elem: K) : Some[V] = {
```

Comment:
Wouldn't it be better to `AnyRef` this and then `Some(nd).asInstanceOf[Option[V]]` in the `remove` code, so the `subtractOne` implementation doesn't have to deal with the superfluous `Some` boxing?

Reply:
`remove0` can return `null` so we either need to box or use a sentinel value for `null`. AFAIR I benchmarked this for the immutable red-black tree and this version performed best.

Reply:
Actually using a sentinel value is a bit faster:
```
Old:
[info] Benchmark                                          (size)  Mode  Cnt       Score      Error  Units
[info] CollisionProofHashMapBenchmark.oaClearRemove          100  avgt   20    2608.536 ±   32.542  ns/op
[info] CollisionProofHashMapBenchmark.oaClearRemove         1000  avgt   20   25750.067 ±  394.784  ns/op
[info] CollisionProofHashMapBenchmark.oaClearRemove        10000  avgt   20  331050.882 ± 6398.876  ns/op
[info] CollisionProofHashMapBenchmark.oaClearSubtractOne     100  avgt   20    2605.293 ±   32.422  ns/op
[info] CollisionProofHashMapBenchmark.oaClearSubtractOne    1000  avgt   20   26160.560 ±  272.107  ns/op
[info] CollisionProofHashMapBenchmark.oaClearSubtractOne   10000  avgt   20  330865.468 ± 5699.215  ns/op

New:
[info] Benchmark                                          (size)  Mode  Cnt       Score      Error  Units
[info] CollisionProofHashMapBenchmark.oaClearRemove          100  avgt   20    2596.609 ±   40.714  ns/op
[info] CollisionProofHashMapBenchmark.oaClearRemove         1000  avgt   20   25700.461 ±  373.438  ns/op
[info] CollisionProofHashMapBenchmark.oaClearRemove        10000  avgt   20  322283.111 ± 4590.289  ns/op
[info] CollisionProofHashMapBenchmark.oaClearSubtractOne     100  avgt   20    2373.294 ±   24.814  ns/op
[info] CollisionProofHashMapBenchmark.oaClearSubtractOne    1000  avgt   20   26747.311 ±  340.112  ns/op
[info] CollisionProofHashMapBenchmark.oaClearSubtractOne   10000  avgt   20  307834.577 ± 4173.648  ns/op
```